OpenRA.Mods.Common/Activities/Move/Move.cs

```diff
@@ -150,8 +150,38 @@ protected override void OnFirstRun(Actor self)
 			}
 		}
 
+		public override bool Cancel(Actor self, bool keepQueue = false)
+		{
+			// Let child activities do what they want so that units don't end up in strange state.
+			// Just clear the path so that child activity doesn't run for too long.
+			// Although we always return false here, Move with nowhere to move will eventually
+			// proceed to the NextActivity when the actor is in good position.
+			if (ChildActivity == null)
+				return base.Cancel(self, keepQueue);
+
+			if (path != null)
+				path.Clear();
+
+			if (keepQueue == false && NextActivity != null)
+				NextActivity.Cancel(self, keepQueue);
+
+			// Could be turn activity, which should be canceled.
+			// To prevent Move first/second half activities from being interrupted,
+			// we set IsInterruptible to false in their constructor.
+			return ChildActivity.Cancel(self, false);
+		}
+
 		public override Activity Tick(Actor self)
 		{
+			// ChildActivity is the top priority, unlike other activities.
+			// Even if this activity is canceled, we must let the child be run so that units
+			// will not end up in an odd place.
+			if (ChildActivity != null)
+			{
+				ActivityUtils.RunActivity(self, ChildActivity);
```

Comment:
> Shouldn't this be assigning the returned activity back to ChildActivity?

You can, but it's not necessary. As forcecore found out, `OuterTick` automatically makes sure that the parent's `ChildActivity` pointer gets updated.

> I think it is wrong to do implicit management here

It has to be done there, as there is no other place where that could happen for a `CompositeActivity`.  Your example shouldn't matter at all, as neither `ChildActivity` nor `Tick`s return value are used there. But let's discuss that in the other PR.